Skip unnecessary loading of `IndexMetadata` during snapshot deletion

joshua-adams-1 opened the pull request:
Skip unnecessary loading of `IndexMetadata` during snapshot deletion

During snapshot deletion we load the metadata for an index into heap purely to calculate the shard count. On nodes with small heaps, loading multiple objects concurrently can cause them to OOMe.

Rather than recomputing the shard count value for the same index across multiple snapshots, this change stores the shard count in a map. This stops loading multiple IndexMetaData objects into heap for the same index.

This means that for commands such as:
```
assertAcked(client().admin().cluster().prepareDeleteSnapshot(timeout, repoName, snapshotName1, snapshotName2).get())
```
assuming both snapshots include Index1, we only loaded Index1's metadata once.

For commands such as:
```
assertAcked(client().admin().cluster().prepareDeleteSnapshot(timeout, repoName, snapshotName1).get())
assertAcked(client().admin().cluster().prepareDeleteSnapshot(timeout, repoName, snapshotName2).get())
```
since these are different delete snapshot requests, we load Index1's metadata once for each

Relates to: https://github.com/elastic/elasticsearch/issues/131822

Closes ES-12539

<!--

---

As a note, I did consider an alternative approach. 

We currently load the index metadata [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java#L1855) inside the `finalizeSnapshot` method. Since the number of primary shards does not change for an index, we could store the shard count here instead (moving the `indexUUIDToShardCountMap` up to the `BlobStoreRepository` level rather than `SnapshotDeletion`). Then, when deleting snapshots, the `IndexMetaData` would not need to be loaded at all.

There were however two issues with this:
1. The `indexUUIDToShardCountMap` map would grow every time an index was added, so clusters adding thousands of indices would quickly grow this map, and there would need to be some form of clean up logic added.
2. If the repository is deleted, then this map does not exist. Therefore I would still need to modify the `determineShardCount` as shown in this PR. 

The benefit is that both commands
```
assertAcked(client().admin().cluster().prepareDeleteSnapshot(timeout, repoName, snapshotName1, snapshotName2).get())
assertAcked(client().admin().cluster().prepareDeleteSnapshot(timeout, repoName, snapshotName2).get())
```
would not load *any* index metadata into heap, assuming the repository was not deleted.

This would also implicitly close ES-12538 since we would no longer have any concurrency concerns
-->

joshua-adams-1 committed:
Skip unnecessary loading of `IndexMetadata` during snapshot deletion

During snapshot deletion we load the metadata for an index into
heap purely to calculate the shard count. Since the number of
primary shards does not change for an index, we can store this to
avoid recomputation, and avoid repeatedly loading large metadata
objects into heap which can cause small nodes to OOMe.

Closes ES-12539

joshua-adams-1 committed:
Skip unnecessary loading of `IndexMetadata` during snapshot deletion

During snapshot deletion we load the metadata for an index into
heap purely to calculate the shard count. Since the number of
primary shards does not change for an index, we can store this when the
snapshot is being created, so that upon deletion we avoid recomputing
this value and loading large metadata objects into heap which can cause
small nodes to OOMe.

Closes ES-12539
Closes ES-12538
Closes #100569

Commit:
[CI] Auto commit changes from spotless

joshua-adams-1 committed:
Skip unnecessary loading of `IndexMetadata` during snapshot deletion

During snapshot deletion we load the metadata for an index into
heap purely to calculate the shard count. On nodes with small heaps,
loading multiple objects concurrently causes them to OOMe.

Rather than recomputing the shard count value for the same index across
multiple snapshots, this change stores the shard count in a map.
This stops loading multiple IndexMetaData objects for the same index
into memory.

Closes ES-12539

joshua-adams-1 committed:
Merge branch 'unnecessary-loading-index-metadata' of github.com:joshua-adams-1/elasticsearch into unnecessary-loading-index-metadata

joshua-adams-1 commented on server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java:
```diff
@@ -183,14 +186,21 @@ public String toString() {
-            + indexMetaData.getMappingVersion()
-            + "-"
-            + indexMetaData.getAliasesVersion();
+        return indexMetaData.getIndexUUID() + DELIMITER + indexMetaData.getSettings()
```
I prefer the other format but spotless keeps reformatting it to this

joshua-adams-1 committed:
Merge branch 'main' into unnecessary-loading-index-metadata

DaveCTurner commented:
> would not load any index metadata into heap, assuming the repository was not deleted.

I think the safest way to do this would be to record the (max) shard count for each index in the `RepositoryData` stored in the repository. We can't reasonably keep it in memory on the master, updating it as we finalize snapshots, because then it'd be lost on a master failover.

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java:
```diff
@@ -174,6 +174,9 @@ public String toString() {
         return "IndexMetaDataGenerations{lookup:" + lookup + "}{identifier:" + identifiers + "}";
     }
 
+    // Cannot use - as a delimiter because getIndexUUID() produces UUIDs with - in them
+    private static final String DELIMITER = "/";
```
Yes exactly.

joshua-adams-1 committed:
Moves indexUUIDToShardCountMap into IndexSnapshotsDeletion

joshua-adams-1 committed:
Refactors determineShardCount

Adds extra test capability to handle deleted indices

joshua-adams-1 committed:
Extend tests to delete and recreate indices

joshua-adams-1 committed:
Merge branch 'unnecessary-loading-index-metadata' of https://github.com/joshua-adams-1/elasticsearch into unnecessary-loading-index-metadata

joshua-adams-1 committed:
Merge branch 'main' into unnecessary-loading-index-metadata

joshua-adams-1 committed:
Fix comments

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1284,37 +1290,47 @@ void run(ActionListener<Void> listener) {
-                        // unnecessary to read multiple metadata blobs corresponding to the same index UUID.
-                        // TODO Skip this unnecessary work? Maybe track the shard count in RepositoryData?
-                        snapshotExecutor.execute(ActionRunnable.run(listeners.acquire(), () -> getOneShardCount(indexMetaGeneration)));
+                        snapshotExecutor.execute(ActionRunnable.run(listeners.acquire(), () -> {
```
I'd rather we didn't fork the no-op tasks - can we check the index UUID first?

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1284,37 +1290,47 @@ void run(ActionListener<Void> listener) {
+                            // index UUID; the shard count is going to be the same for all metadata with the same index UUID, so it is
+                            // unnecessary to read multiple metadata blobs corresponding to the same index UUID.
+                            String indexUUID = originalRepositoryData.indexMetaDataGenerations().convertBlobIdToIndexUUID(blobId);
+                            if (indexUUIDToShardCountMap.containsKey(indexUUID) == false) {
```
This still risks duplicates since we call `containsKey` separately from `put`. Once we're just tracking the set of the index UUIDs we can just call `add` here which returns a flag indicating whether the item was newly added.

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1284,37 +1290,47 @@ void run(ActionListener<Void> listener) {
+                            // NB since 7.9.0 we deduplicate index metadata blobs, and one of the components of the deduplication key is the
+                            // index UUID; the shard count is going to be the same for all metadata with the same index UUID, so it is
+                            // unnecessary to read multiple metadata blobs corresponding to the same index UUID.
+                            String indexUUID = originalRepositoryData.indexMetaDataGenerations().convertBlobIdToIndexUUID(blobId);
```
This will be `null` if the index metadata blob is in the pre-7.9.0 format, but we have to load all such blobs. Not sure if the map accepts a `null` key, but even if it does it won't accept more than one of them.

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java:
```diff
@@ -193,4 +196,34 @@ public static String buildUniqueIdentifier(IndexMetadata indexMetaData) {
+        // The uniqueIdentifier was built in buildUniqueIdentifier, is of the format IndexUUID-String-long-long-long,
+        // and uses '-' as a delimiter.
+        // The below regex accounts for the fact that the IndexUUID can also contain the '-' character
+        Pattern pattern = Pattern.compile("^(.*?)-[^-]+-\\d+-\\d+-\\d+$");
```
<img width="720" height="479" alt="Image" src="https://github.com/user-attachments/assets/b99cd14b-a565-420e-b31d-3a885880c40b" />

The index UUID (and the history UUID) are fixed-length. Can we just split the string at the appropriate index?

joshua-adams-1 replied:
AFAICT the length is set to 22, [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/UUIDs.java#L63), or set to `_na_` if unknown, [here](https://github.com/elastic/elasticsearch/blob/9435f4104c8ea56bcbd1a52a4ec0d70baf26f283/server/src/main/java/org/elasticsearch/cluster/ClusterState.java#L158). Will simplify this accordingly

DaveCTurner reviewed (commented):
Some comments on the production code. I skimmed through the tests and they look to be doing the right sort of thing, will come back to them in a later cycle.

joshua-adams-1 committed:
David comments

joshua-adams-1 committed:
Clean up

joshua-adams-1 committed:
Further clea up:

- Remove unreferenced method
- Rename convertBlobIdToIndexUUID to getIndexUUIDFromBlobId
- Tweak comments

joshua-adams-1 committed:
Refactor Tests

joshua-adams-1 committed:
Merge branch 'main' into unnecessary-loading-index-metadata

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java:
```diff
@@ -193,4 +211,30 @@ public static String buildUniqueIdentifier(IndexMetadata indexMetaData) {
+        // The indexUUID is either a UUID of length 22, or _na_
+        boolean na = uniqueIdentifier.startsWith(ClusterState.UNKNOWN_UUID + "-");
+        if (na) {
+            return ClusterState.UNKNOWN_UUID;
```
I don't think this is possible for a blob that actually ended up in the repository - we use `_na_` for the index UUID only if the index hasn't been created yet, but in that case we couldn't possibly have snapshotted it.

DaveCTurner reviewed (commented):
Sorry got to dash but I saw a couple more things. Will give it a more thorough look next week.

joshua-adams-1 committed:
Comments for IndexMetaDataGenerations

joshua-adams-1 committed:
Merge branch 'unnecessary-loading-index-metadata' of https://github.com/joshua-adams-1/elasticsearch into unnecessary-loading-index-metadata

joshua-adams-1 committed:
Merge branch 'main' into unnecessary-loading-index-metadata

joshua-adams-1 committed:
David Comments

joshua-adams-1 committed:
Merge branch 'main' into unnecessary-loading-index-metadata

joshua-adams-1 committed:
Fixing comment and variable name

Commit:
[CI] Auto commit changes from spotless

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1311,37 +1312,47 @@ void run(ActionListener<Void> listener) {
             private void determineShardCount(ActionListener<Void> listener) {
                 try (var listeners = new RefCountingListener(listener)) {
-                    for (final var indexMetaGeneration : snapshotIds.stream()
+                    Map<String, String> blobIdToIndexUuidMap = originalRepositoryData.indexMetaDataGenerations().getBlobIdToIndexUuidMap();
```
This gets called once per index, which could still be (hundreds-of-)thousands of times, but it returns the same value each time. It really needs to be done once for the entire snapshot deletion process.

DaveCTurner reviewed (commented):
Given the massive savings we've been able to find in #137210 I wonder if we should defer this for now. Computing `blobIdToIndexUuidMap` is kinda expensive, even if we do it just once, and its expense scales with the size of the repository rather than the number of indices being deleted - if we're only deleting a few indices then most of that expense is wasted. We can find ways to refine this for sure, and it's kinda nice that this might reduce the number of blobs we read during a large delete, but I don't think this is critical once #137210 lands and I worry that it might end up causing more problems than it solves.

joshua-adams-1 committed:
Move blobIdToIndexUuidMap int snapshots deletion

joshua-adams-1 committed:
Merge branch 'unnecessary-loading-index-metadata' of https://github.com/joshua-adams-1/elasticsearch into unnecessary-loading-index-metadata

joshua-adams-1 committed:
Merge branch 'main' into unnecessary-loading-index-metadata

ywangd commented:
> https://github.com/elastic/elasticsearch/pull/137210 reduces the marginal benefits of this change.

I was wondering whether we still need this PR. My hunch is no. It might be helpful in certain theoretical cases. But still probably better to address them when actually observed?

joshua-adams-1 committed:
Merge branch 'main' into unnecessary-loading-index-metadata
